File: dc_bridge/dc_bridge_core/src/uploader/multipart.rs
//! Multipart, resumable uploads for large Files (ADR-0005): robot networks are flaky
//! and videos/maps are large, so an interrupted transfer must resume, not restart.
//!
//! Built on `object_store`'s low-level [`MultipartStore`] API rather than the
//! high-level `put_multipart` writer, because only the low-level API exposes the
//! persistent [`MultipartId`] and per-part [`PartId`]s that make resumption possible:
//! both are checkpointed to a JSON sidecar file after every part, so a new process (or
//! a retry after a dropped connection) re-reads the sidecar, skips every part that
//! already landed, and finishes the same multipart upload instead of starting over.
//! The sidecar is removed once the upload completes.

use object_store::multipart::PartId;
use object_store::path::Path as ObjectPath;
use object_store::{MultipartId, PutPayload};
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{Read, Seek, SeekFrom};
use std::path::{Path, PathBuf};

use super::store::UploadStore;

/// Progress of one interrupted multipart upload, checkpointed after every part.
#[derive(Debug, Serialize, Deserialize)]
struct ResumeState {
    multipart_id: MultipartId,
    part_size: u64,
    file_size: u64,
    /// `content_id` of each completed part, by part index; `None` = not yet uploaded.
    parts: Vec<Option<String>>,
}

/// Where the resume sidecar for one (storage, object key) upload lives. Stable across
/// runs — FNV-1a rather than `DefaultHasher`, whose keys are randomized per process.
pub fn resume_state_path(state_dir: &Path, storage_name: &str, object_key: &str) -> PathBuf {
    let mut hash: u64 = 0xcbf2_9ce4_8422_2325;
    for byte in storage_name.bytes().chain([0u8]).chain(object_key.bytes()) {
        hash ^= u64::from(byte);
        hash = hash.wrapping_mul(0x0000_0100_0000_01b3);
    }
    state_dir.join(format!("multipart-{hash:016x}.json"))
}

fn generic_err(
    source: impl Into<Box<dyn std::error::Error + Send + Sync + 'static>>,
) -> object_store::Error {
    object_store::Error::Generic {
        store: "dc_bridge uploader resume state",
        source: source.into(),
    }
}

fn load_state(state_path: &Path, part_size: u64, file_size: u64) -> Option<ResumeState> {
    let content = fs::read_to_string(state_path).ok()?;
    let state: ResumeState = serde_json::from_str(&content).ok()?;
    // A sidecar from a different policy or a changed file can't be resumed safely.
    (state.part_size == part_size
        && state.file_size == file_size
        && state.parts.len() == part_count(file_size, part_size))
    .then_some(state)
}

fn save_state(state_path: &Path, state: &ResumeState) -> Result<(), object_store::Error> {
    let tmp = state_path.with_extension("json.tmp");
    let content = serde_json::to_string(state).map_err(generic_err)?;
    fs::write(&tmp, content).map_err(generic_err)?;
    fs::rename(&tmp, state_path).map_err(generic_err)?;
    Ok(())
}

fn part_count(file_size: u64, part_size: u64) -> usize {
    (file_size.div_ceil(part_size)).max(1) as usize
}

/// Uploads `local` to `location` on `store` in `part_size`-byte parts, checkpointing
/// progress to `state_path` so an interrupted upload resumes from the last completed
/// part. `file_size` must be `local`'s current size (the caller already stat'ed it).
pub async fn upload_resumable(
    store: &dyn UploadStore,
    location: &ObjectPath,
    local: &Path,
    file_size: u64,
    part_size: u64,
    state_path: &Path,
) -> Result<(), object_store::Error> {
    let n_parts = part_count(file_size, part_size);

    let mut state = match load_state(state_path, part_size, file_size) {
        Some(state) => state,
        None => {
            let multipart_id = store.create_multipart(location).await?;
            let state = ResumeState {
                multipart_id,
                part_size,
                file_size,
                parts: vec![None; n_parts],
            };
            save_state(state_path, &state)?;
            state
        }
    };

    let mut file = fs::File::open(local).map_err(generic_err)?;
    for idx in 0..n_parts {
        if state.parts[idx].is_some() {
            continue;
        }
        let offset = idx as u64 * part_size;
        let len = part_size.min(file_size - offset) as usize;
        let mut buf = vec![0u8; len];
        file.seek(SeekFrom::Start(offset)).map_err(generic_err)?;
        file.read_exact(&mut buf).map_err(generic_err)?;

        let part = store
            .put_part(location, &state.multipart_id, idx, PutPayload::from(buf))
            .await?;
        state.parts[idx] = Some(part.content_id);
        save_state(state_path, &state)?;
    }

    let parts: Vec<PartId> = state
        .parts
        .iter()
        .map(|content_id| PartId {
            content_id: content_id
                .clone()
                .expect("every part index was just filled in"),
        })
        .collect();
    store
        .complete_multipart(location, &state.multipart_id, parts)
        .await?;
    let _ = fs::remove_file(state_path);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resume_state_paths_are_stable_and_distinct() {
        let dir = Path::new("/state");
        let a = resume_state_path(dir, "minio", "maps/map.pgm");
        assert_eq!(a, resume_state_path(dir, "minio", "maps/map.pgm"));
        assert_ne!(a, resume_state_path(dir, "s3", "maps/map.pgm"));
        assert_ne!(a, resume_state_path(dir, "minio", "maps/other.pgm"));
    }

    #[test]
    fn part_count_covers_edge_sizes() {
        assert_eq!(part_count(0, 1024), 1);
        assert_eq!(part_count(1, 1024), 1);
        assert_eq!(part_count(1024, 1024), 1);
        assert_eq!(part_count(1025, 1024), 2);
    }
}

File: dc_bridge/src/uploader/multipart.cpp
// SPDX-FileCopyrightText: 2022-2026 David Bensoussan
// SPDX-License-Identifier: MPL-2.0

#include "dc_bridge/uploader/multipart.hpp"

#include <cstdio>
#include <fstream>
#include <nlohmann/json.hpp>
#include <optional>
#include <vector>

namespace dc_bridge::uploader::multipart
{

namespace
{

std::size_t part_count(std::uint64_t file_size, std::uint64_t part_size)
{
  std::uint64_t n = (file_size + part_size - 1) / part_size;  // ceil
  return static_cast<std::size_t>(n < 1 ? 1 : n);
}

struct ResumeState
{
  std::string upload_id;
  std::uint64_t part_size;
  std::uint64_t file_size;
  std::vector<std::optional<std::string>> parts;  // ETag per part index; nullopt = pending.
};

// A sidecar from a different policy or a changed file can't be resumed safely.
std::optional<ResumeState> load_state(const std::string& state_path, std::uint64_t part_size, std::uint64_t file_size)
{
  std::ifstream in(state_path);
  if (!in.good())
  {
    return std::nullopt;
  }
  nlohmann::json j;
  try
  {
    in >> j;
  }
  catch (const nlohmann::json::exception&)
  {
    return std::nullopt;
  }
  ResumeState s;
  try
  {
    s.upload_id = j.at("upload_id").get<std::string>();
    s.part_size = j.at("part_size").get<std::uint64_t>();
    s.file_size = j.at("file_size").get<std::uint64_t>();
    for (const auto& p : j.at("parts"))
    {
      s.parts.push_back(p.is_null() ? std::nullopt : std::optional<std::string>(p.get<std::string>()));
    }
  }
  catch (const nlohmann::json::exception&)
  {
    return std::nullopt;
  }
  if (s.part_size != part_size || s.file_size != file_size || s.parts.size() != part_count(file_size, part_size))
  {
    return std::nullopt;
  }
  return s;
}

void save_state(const std::string& state_path, const ResumeState& s)
{
  nlohmann::json j;
  j["upload_id"] = s.upload_id;
  j["part_size"] = s.part_size;
  j["file_size"] = s.file_size;
  nlohmann::json parts = nlohmann::json::array();
  for (const auto& p : s.parts)
  {
    parts.push_back(p ? nlohmann::json(*p) : nlohmann::json(nullptr));
  }
  j["parts"] = std::move(parts);

  const std::string tmp = state_path + ".tmp";
  {
    std::ofstream out(tmp, std::ios::trunc);
    out << j.dump();
    if (!out.good())
    {
      throw ObjectStoreError("failed to write multipart resume sidecar " + tmp);
    }
  }
  if (std::rename(tmp.c_str(), state_path.c_str()) != 0)
  {
    throw ObjectStoreError("failed to rename multipart resume sidecar into place");
  }
}

}  // namespace

std::string resume_state_path(const std::string& state_dir, const std::string& storage_name,
                              const std::string& object_key)
{
  // FNV-1a over storage_name, a NUL, then object_key.
  std::uint64_t hash = 0xcbf29ce484222325ULL;
  auto mix = [&hash](unsigned char byte) {
    hash ^= static_cast<std::uint64_t>(byte);
    hash *= 0x100000001b3ULL;
  };
  for (unsigned char c : storage_name)
  {
    mix(c);
  }
  mix(0);
  for (unsigned char c : object_key)
  {
    mix(c);
  }
  char name[32];
  std::snprintf(name, sizeof(name), "multipart-%016lx.json", static_cast<unsigned long>(hash));
  std::string sep = (!state_dir.empty() && state_dir.back() == '/') ? "" : "/";
  return state_dir + sep + name;
}

void upload_resumable(ObjectStore& store, const std::string& key, const std::string& local_path,
                      std::uint64_t file_size, std::uint64_t part_size, const std::string& state_path)
{
  const std::size_t n_parts = part_count(file_size, part_size);

  std::optional<ResumeState> loaded = load_state(state_path, part_size, file_size);
  ResumeState state;
  if (loaded)
  {
    state = std::move(*loaded);
  }
  else
  {
    state.upload_id = store.create_multipart(key);
    state.part_size = part_size;
    state.file_size = file_size;
    state.parts.assign(n_parts, std::nullopt);
    save_state(state_path, state);
  }

  std::ifstream file(local_path, std::ios::binary);
  if (!file.good())
  {
    throw ObjectStoreError("failed to open local file for multipart upload: " + local_path);
  }
  for (std::size_t idx = 0; idx < n_parts; ++idx)
  {
    if (state.parts[idx])
    {
      continue;
    }
    const std::uint64_t offset = static_cast<std::uint64_t>(idx) * part_size;
    const std::size_t len = static_cast<std::size_t>(std::min(part_size, file_size - offset));
    std::string buf(len, '\0');
    file.seekg(static_cast<std::streamoff>(offset));
    file.read(buf.data(), static_cast<std::streamsize>(len));
    if (static_cast<std::size_t>(file.gcount()) != len)
    {
      throw ObjectStoreError("short read while assembling multipart part " + std::to_string(idx));
    }
    // ObjectStore part numbers are 1-based (S3 convention).
    std::string etag = store.put_part(key, state.upload_id, static_cast<int>(idx) + 1, buf);
    state.parts[idx] = etag;
    save_state(state_path, state);
  }

  std::vector<std::string> etags;
  etags.reserve(state.parts.size());
  for (const auto& p : state.parts)
  {
    etags.push_back(*p);  // every part index was just filled in
  }
  store.complete_multipart(key, state.upload_id, etags);
  std::remove(state_path.c_str());
}

}  // namespace dc_bridge::uploader::multipart
